sdk/testing/stepwise/environments/docker/environment.go

```diff
@@ -0,0 +1,851 @@
+	ClusterName   string
+	RootCAs       *x509.CertPool
+	CACert        *x509.Certificate
+	CAKey         *ecdsa.PrivateKey
+	CleanupFunc   func()
+	SetupFunc     func()
+	ClusterNodes  []*DockerClusterNode
+
+	// networkID tracks the network ID of the created docker network
+	networkID string
+
+	barrierKeys  [][]byte
+	recoveryKeys [][]byte
+	tmpDir       string
+
+	clientAuthRequired bool
+	// the mountpath of the plugin under test
+	mountPath string
+	// rootToken is the initial root token created when the Vault cluster is
+	// created.
+	rootToken string
+}
+
+// Teardown stops all the containers.
+func (rc *DockerCluster) Teardown() error {
+	var result error
+	for _, node := range rc.ClusterNodes {
+		if err := node.Cleanup(); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	//clean up networks
+	if rc.networkID != "" {
+		cli, err := docker.NewClientWithOpts(docker.FromEnv, docker.WithVersion("1.40"))
+		if err != nil {
+			return err
+		}
+		if err := cli.NetworkRemove(context.Background(), rc.networkID); err != nil {
+			return err
+		}
+	}
+
+	return result
+}
+
+// MountPath returns the path that the plugin under test is mounted at. If a
+// MountPathPrefix was given, the mount path uses the prefix with a uuid
+// appended. The default is the given PluginName with a uuid suffix.
+func (dc *DockerCluster) MountPath() string {
+	if dc.mountPath != "" {
+		return dc.mountPath
+	}
+
+	uuidStr, _ := uuid.GenerateUUID()
+	prefix := dc.PluginName
+	if dc.MountOptions.MountPathPrefix != "" {
+		prefix = dc.MountOptions.MountPathPrefix
+	}
+
+	dc.mountPath = fmt.Sprintf("%s_%s", prefix, uuidStr)
+	if dc.MountOptions.PluginType == stepwise.PluginTypeCredential {
+		dc.mountPath = fmt.Sprintf("%s/%s", "auth", dc.mountPath)
+	}
+
+	return dc.mountPath
+}
+
+// RootToken returns the root token of the cluster, if set
+func (dc *DockerCluster) RootToken() string {
+	return dc.rootToken
+}
+
+// Name returns the name of this environment
+func (dc *DockerCluster) Name() string {
+	return "Docker"
+}
+
+// Client returns a clone of the configured Vault API client.
+func (dc *DockerCluster) Client() (*api.Client, error) {
+	if len(dc.ClusterNodes) > 0 {
+		if dc.ClusterNodes[0].Client != nil {
+			c, err := dc.ClusterNodes[0].Client.Clone()
+			if err != nil {
+				return nil, err
+			}
+			c.SetToken(dc.ClusterNodes[0].Client.Token())
+			return c, nil
+		}
+	}
+
+	return nil, errors.New("no configured client found")
+}
+
+func (n *DockerClusterNode) Name() string {
+	return n.Cluster.ClusterName + "-" + n.NodeID
+}
+
+func (rc *DockerCluster) Initialize(ctx context.Context) error {
+	client, err := rc.ClusterNodes[0].CreateAPIClient()
+	if err != nil {
+		return err
+	}
+
+	var resp *api.InitResponse
+	for ctx.Err() == nil {
+		resp, err = client.Sys().Init(&api.InitRequest{
+			SecretShares:    3,
+			SecretThreshold: 3,
+		})
+		if err == nil && resp != nil {
+			break
+		}
+		time.Sleep(500 * time.Millisecond)
+	}
```

Comment:
You could use `time.NewTicker` to avoid resetting the timer.

Reply:
A ticker wouldn't account for the time the request is happening though.

Reply:
Deferring and tracking this internally, to follow up in the coming weeks

Reply:
> NewTicker returns a new Ticker containing a channel that will send the time with a period specified by the duration argument. It adjusts the intervals or drops ticks to make up for slow receivers.

Using a ticker should be fine in this case, since it's being used to control the request rate. If the requests takes shorter than the tick duration it will block until then, and if it's longer the ticker simply drops the tick and allows the next request to be processed immediately.